contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -203,24 +238,34 @@ namespace eosiosystem {
       // Check if the finalizer key is not already active
       check( finalizer_key_itr->id != finalizer->active_finalizer_key_id, "finalizer key was already active: " + finalizer_key );
 
-      auto prev_fin_key_id = finalizer->active_finalizer_key_id;
+      auto active_key_id = finalizer->active_finalizer_key_id;
 
+      // Update finalizer's information in _finalizers table
       _finalizers.modify( finalizer, same_payer, [&]( auto& f ) {
          f.active_finalizer_key_id      = finalizer_key_itr->id;
          f.active_finalizer_key_binary  = finalizer_key_itr->finalizer_key_binary;
       });
 
-      // Do a binary search to see if the finalizer is in last proposed policy
-      auto itr = std::lower_bound(_gstate_a.last_proposed_keys.begin(), _gstate_a.last_proposed_keys.end(), prev_fin_key_id, [](const proposed_finalizer_key_t& key, uint64_t id) {
+      auto last_proposed_finalizers = get_last_proposed_finalizers();
+      if( last_proposed_finalizers.empty() ) {
+         return;
+      }
+
+      // Search last_proposed_finalizers for active_key_id
+      auto itr = std::lower_bound(last_proposed_finalizers.begin(), last_proposed_finalizers.end(), active_key_id, [](const finalizer_auth_info& key, uint64_t id) {
          return key.key_id < id;
       });
 
-      if( itr != _gstate_a.last_proposed_keys.end() && itr->key_id == prev_fin_key_id ) {
-         // Replace the previous finalizer key with finalizer key just activated
+      // If active_key_id is in last_proposed_finalizers, it means the finalizer is
+      // active. Replace the existing entry in last_proposed_finalizers with
+      // the information of finalizer_key just activated and call set_proposed_finalizers
+      if( itr != last_proposed_finalizers.end() && itr->key_id == active_key_id ) {
+         // Update last_proposed_finalizers
```

Comment:
And only here, when you already know you will need to call `set_proposed_finalizers`, would you pay the cost of doing a copy of the vector and then modify that vector and call `set_proposed_finalizers` with it:

```cpp
auto proposed_finalizers = last_proposed_finalizers;
auto& matching_entry = proposed_finalizers[itr - last_proposed_finalizers.begin()];

matching_entry.key_id = finalizer_key_itr->id;
matching_entry.fin_authority.public_key = finalizer_key_itr->finalizer_key_binary;

set_proposed_finalizers(std::move(proposed_finalizers));
```